File: Sems_2/Pertemuan1/binary.py
import random
import time


def binary_search(data_list, x):
    low = 0
    high = len(data_list) - 1
    mid = 0

    while low <= high:
        mid = (high + low) // 2

        if data_list[mid][0] < x:
            low = mid + 1
        elif data_list[mid][0] > x:
            high = mid - 1
        else:
            return mid
    return -1


# List nama-nama
with open('D:\\Alprot\\Sems_2\\Pertemuan1\\databases.txt', 'r') as file:
    content = file.read()
    names = content.replace("'", "").replace(",", "").split()

# list nilai
with open('D:\\Alprot\\Sems_2\\Pertemuan1\\random_numbers.txt', 'r') as file:
    values = file.read().split()

# Menyusun data dengan nilai
data_nilai = list(zip(names, values))

# Sort the data for binary search
data_nilai.sort(key=lambda x: x[0])

print("Pertemuan 1 - Binary Search in python\n")

x = input("Cari data = ")  # Keep input as string for name search

start_time = time.perf_counter()

hasil = binary_search(data_nilai, x)

end_time = time.perf_counter()
waktu = end_time - start_time

if hasil != -1:
    print(f"Elemen '{x}' ditemukan pada index {
          hasil} dengan nilai {data_nilai[hasil][1]}")
    print("Waktu yang dibutuhkan: {:.8f} detik".format(waktu))
else:
    print(f"Elemen '{x}' tidak ada dalam array")

'''contoh perbandingannya'''
# Elemen 'yuni' ditemukan pada index 998 dengan nilai 51,
# Waktu yang dibutuhkan: 0.00002700 detik

File: Sems_2/Pertemuan1/binary.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <chrono>
#include <algorithm>

using namespace std;
using namespace std::chrono;

int binary_search(vector<pair<string, string>> data_list, string x)
{
    int l = 0, r = data_list.size() - 1;
    while (l <= r)
    {
        int m = l + (r - l) / 2;
        if (data_list[m].first == x)
            return m;
        if (data_list[m].first < x)
            l = m + 1;
        else
            r = m - 1;
    }
    return -1;
}

int main()
{
    ifstream names_file("D:\\Alprot\\Sems_2\\Pertemuan1\\new_databases.txt");
    ifstream values_file("D:\\Alprot\\Sems_2\\Pertemuan1\\new_numbers.txt");

    vector<string> names;
    vector<string> values;
    string line;

    while (getline(names_file, line))
    {
        names.push_back(line);
    }

    while (getline(values_file, line))
    {
        values.push_back(line);
    }

    vector<pair<string, string>> data_nilai;
    for (int i = 0; i < names.size(); i++)
    {
        data_nilai.push_back(make_pair(names[i], values[i]));
    }

    sort(data_nilai.begin(), data_nilai.end());


    cout << "Pertemuan 1 - Binary Search in cpp" << endl;
    string x;
    cout << "Cari data = ";
    cin >> x;

    auto start_time = high_resolution_clock::now();

    int hasil = binary_search(data_nilai, x);

    auto end_time = high_resolution_clock::now();
    auto waktu = duration_cast<nanoseconds>(end_time - start_time).count();

    if (hasil != -1)
    {
        cout << "Elemen '" << x << "' ditemukan pada index " << hasil << " dengan nilai " << data_nilai[hasil].second << endl;
        cout << "Waktu yang dibutuhkan: " << waktu << " nanoseconds" << endl;
    }
    else
    {
        cout << "Elemen '" << x << "' tidak ada dalam array" << endl;
    }

    return 0;
}